### PR TITLE
 [BUG] Refactor ADICVTransformer and fix CV calculation

### LINKED ISSUE #6755: [BUG] CV calculation is incorrect in ADICVTransformer class
**Describe the bug**
Coefficient of Variation (CV) should be calculated as the sigma of non-zero values squared over the mu of non-zero values squared; however, in the implementation here, it's being calculated as the variance of non-zero values (Which should correctly correspond to sigma squared) over the number of non-zero values, making the denominator incorrect.

This blog post contains a short description of the calculations and approach: https://frepple.com/blog/demand-classification/

Source code reference:
https://github.com/sktime/sktime/blob/cd769b1769be67888213bfff522e1ea86f71eaef/sktime/transformations/series/adi_cv.py#L167

Discord reference:
https://discord.com/channels/1075852648688930887/1075852648948973642/1261316663991861273

**To Reproduce**
Following the blog post example, here's the code to produce an incorrect ADI, CV, and Classification:

```python
import numpy as np
import pandas as pd
from datetime import datetime
from sktime.transformations.series.adi_cv import ADICVTransformer

transformer = ADICVTransformer()
periods = pd.date_range(start=datetime(2023, 1, 1), end=datetime(2023, 12, 1), freq='MS')
series = [6, 5, 0, 9, 2, 0, 14, 0, 0, 21, 0, 17]
df = pd.DataFrame({'series': series}, index=periods)

adi_cv = transformer.fit_transform(df)
print(adi_cv)
```

output:
|adi	|cv2	|class|
|:----|:-----|:----|
|2.0	|6.897959	|lumpy

**Expected behavior**
The output should look like this:
|adi	|cv2	|class|
|:----|:-----|:----|
|1.71|0.37	|intermittent

**Additional context**

**Versions**
[c:\Users\sbhobbes\Anaconda3\envs\apf\Lib\site-packages\statsforecast\core.py:27](file:///C:/Users/sbhobbes/Anaconda3/envs/apf/Lib/site-packages/statsforecast/core.py:27): TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from tqdm.autonotebook import tqdm

System:
    python: 3.11.7 | packaged by conda-forge | (main, Dec 23 2023, 14:27:59) [MSC v.1937 64 bit (AMD64)]
executable: [c:\Users\sbhobbes\Anaconda3\envs\apf\python.exe](file:///C:/Users/sbhobbes/Anaconda3/envs/apf/python.exe)
   machine: Windows-10-10.0.19042-SP0

Python dependencies:
          pip: 23.3.2
       sktime: 0.30.2
      sklearn: 1.4.0
       skbase: 0.8.1
        numpy: 1.26.3
        scipy: 1.12.0
       pandas: 2.2.2
   matplotlib: 3.8.2
       joblib: 1.3.2
        numba: 0.59.1
  statsmodels: 0.14.1
     pmdarima: 2.0.4
statsforecast: 1.7.5
      tsfresh: None
      tslearn: None
        torch: 2.3.1+cpu
   tensorflow: None
tensorflow_probability: None

</details>

### PR DESCRIPTION
**Pull request by sbhobbes**
Fixes coefficient of variation calculation and refactors some of the corresponding classification code. Also adds parameter for the method used to calculate ADI. Updates tests to remove randomly-generated test data and create reproducible tests instead.

#### Reference Issues/PRs
Fixes #6755 

#### What does this implement/fix? Explain your changes.
Addresses an incorrect denominator in the coefficient of variance squared calculation.  It also adds a parameter to allow the end user to control the end behavior of the average demand interval calculation.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

Please review the functionality and accuracy of the updated calculations for the coefficient of variance squared and the parameterized average demand interval; focusing specifically on opportunities for errors/exceptions (e.g. division by zero error).  Also need updates to the documentation and unit tests.

#### Did you add any tests for the change?

Yes, original tests have been altered; additional tests are in development.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by sbhobbes** [comment]
@all-contributors please add @sbhobbes for <code>

**Comment by sbhobbes** [comment]
@all-contributors please add @sbhobbes for :code:

**Comment by sbhobbes** [comment]
@all-contributors please add @sbhobbes for code

File: sktime/transformations/series/adi_cv.py [hunk 1]
```diff
@@ -15,27 +16,27 @@ class ADICVTransformer(BaseTransformer):
...
 
     Let :math:`x_t` be the value of the time series at times :math:`t = 1, 2, \dots, T`,
     and let :math:`N` be the number of non-zero values in the time series, i.e.,
     :math:`N = \text{card}\{t: x_t \neq 0\}`.
 
-    The ADI and CV2 are calculated as follows:
+    The ADI and CV^2 are calculated as follows:
 
-    1. Average Demand Interval (ADI): The average time period between
-      time periods with non-zero demands, mathematically defined as:
```
Comment (fkiraly) [inline sktime/transformations/series/adi_cv.py:28]: it's actually non-zero, given the code. In-principle, this could be applied to series with negative values, too.

File: sktime/transformations/series/adi_cv.py [hunk 2]
```diff
@@ -59,6 +63,14 @@ class ADICVTransformer(BaseTransformer):
     features : list[str] | None (default = ['adi', 'cv2', 'class'])
         Specifies all of the feature values to be calculated
 
+    adi_trim_handling: string (default = pool)
+        Specifies the method for reconciling leading/trailing zeros in the series.
+        Allowable values are ('pool', 'trim', and 'ignore'), corresponding to the
+        following treatment of leading/trailing zeros:
+        pool => N Observations / N Non-Zero Observations
```
Comment (fkiraly) [inline sktime/transformations/series/adi_cv.py:70]: use of the letter "N" is confusing, since a reader might think it corresponds to the same integer. How about you define the L and N as in the issue to avoid ambiguity?

File: sktime/transformations/series/adi_cv.py [hunk 3]
```diff
@@ -47,6 +48,9 @@ class ADICVTransformer(BaseTransformer):
     Default values for the thresholds are taken from the paper by Syntetos/Boylan [1].
     namely, ``adi_threshold = 1.32`` and ``cv2_threshold = 0.49``.
     They can also be adjusted by passing them as parameters to the transformer.
+    Note that Hyndman & Kostenko [2], in their response to Syntetos/Boylan, indicated that
+    there was a more robust threshold for ADI/CV; this can theoretically be accounted for
```
Comment (fkiraly) [inline sktime/transformations/series/adi_cv.py:52]: the "how" might be useful to tell a reader, if you already cite [2]

Comment (fkiraly) [reply sktime/transformations/series/adi_cv.py:52]: ok, I just looked, that's a pretty confusing paper. It probably assumes context from multiple references therein.

File: sktime/transformations/series/adi_cv.py [hunk 4]
```diff
@@ -153,51 +178,61 @@ def _transform(self, X, y=None):
...
+            elif cv2_value >= self.cv_threshold:
+                class_type = "erratic"
+        elif adi_value >= self.adi_threshold:
+            if cv2_value < self.cv_threshold:
+                class_type = "intermittent"
+            elif cv2_value >= self.cv_threshold:
+                class_type = "lumpy"
 
         # Collecting all values together into dict and converting to DF
-        return_dict = {}
```
Comment (fkiraly) [inline sktime/transformations/series/adi_cv.py:189]: I think the original idea of invoking the `pd.DataFrame` constructor only one is more robust. Assignments in `DataFrame` via the brackets can be buggy, imo we should avoid them. I could tell stories about these breaking after `pandas` updates because the coercion logic changes.

File: sktime/transformations/series/tests/test_adi_cv.py [hunk 5]
```diff
@@ -9,124 +9,56 @@
 from sktime.transformations.series.adi_cv import ADICVTransformer
 
 
-def _generate_smooth_series(size: int = 750):
```
Comment (fkiraly) [inline sktime/transformations/series/tests/test_adi_cv.py:12]: I would leave both tests in, and the generate functions, imo they are useful if we need mor egeneral tests later on.
Perhaps allow the `_generate` functions to set a random seed, or there is a stochatsic failure chance.

**Review by fkiraly — CHANGES_REQUESTED** [review]
thx! left some comments

**Comment by sbhobbes** [comment]
@shlok191 I would appreciate any feedback/thoughts you have here as well.  Thanks!

File: sktime/transformations/series/tests/test_adi_cv.py (hunk 5 above)
Comment (shlok191) [reply sktime/transformations/series/tests/test_adi_cv.py:12]: I agree, perhaps we can specify the `adi_trim_handling` parameter's value to keep the old mathematical formula?

File: sktime/transformations/series/adi_cv.py (hunk 2 above)
Comment (shlok191) [reply sktime/transformations/series/adi_cv.py]: I saw the definitions for L and N. It definitely makes more sense now!

**Comment by shlok191** [comment]
@sbhobbes, looks good to me. Thanks a lot again. I remember being confused about the exact ADI definition as well...There really is no 1 agreed upon definition I could find online. I think providing folks with multiple options is really helpful! :)

**Review by fkiraly — APPROVED** [review]
Great!

The formulae seem to be correct, it would be great if you could double check, @shlok191.

Also, I see the PR is still a draft - do you intent to make further changes?

**Comment by shlok191** [comment]
Everything looks good to me too! @fkiraly

**Comment by sbhobbes** [comment]
@fkiraly Yeah, I'd like to make additional unit tests related to the new parameter for adi_trim_handling, I just got pulled onto a different project this week at work to help another team meet a deadline, so I haven't had a chance to get back into it. I should have something more added by the end of the weekend. 

If you want to pull it out of draft before that to get the fix into production, I'm OK with that and will do a new PR for the additional tests. 

Thank you both again!

**Comment by fkiraly** [comment]
the next release is later than the weekend, so no need to rush for now.

**Comment by sbhobbes** [comment]
> the next release is later than the weekend, so no need to rush for now - weekend is fine!

I added 2 parametrized unit tests for different expected values based on the different calculations.  I'm good with the changes, but please feel free to let me know if you or @shlok191 want me to do anything else.  Thanks again!

File: sktime/transformations/series/adi_cv.py [hunk 6]
```diff
@@ -153,37 +174,49 @@ def _transform(self, X, y=None):
...
-        cv2_value = variance / len(X_non_zero)
+        if self.adi_trim_handling == "trim":
+            first_non_zero = X_non_zero_indices[0]
+            last_non_zero = X_non_zero_indices[-1]
 
-        # Calculating the class type
+            adi_numerator = last_non_zero - first_non_zero
+            adi_denominator = n_non_zero_observations - 1
 
-        adi_low = adi_value <= self.adi_threshold
```
Comment (fkiraly) [inline sktime/transformations/series/adi_cv.py:171]: I would leave the old logic of computing `cv2_low` first - that makes the code easier to read, and debug. May I kindly request to add back `adi_high`, `adi_low`, etc?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

Very minor change request, I would request leaving the `adi_high` and `adi_low` variables, I think that makes it easier to read the code.

The math is very simple, but errors of sign etc are easy to make, so enabling step by step checks by a reviewer makes the code less error prone... we have already seen how easy it is to overlook a bug.

**Review by fkiraly — DISMISSED** [review]
Thanks!
